src/Microsoft.IdentityModel.Tokens/Validation/ValidationResult.cs

```diff
@@ -0,0 +1,110 @@
+    /// A <see cref="TokenValidationResult"/> maintains a list of <see cref="ValidationResult"/> for each step in the token validation.
+    /// </summary>
+    internal abstract class ValidationResult
+    {
+        private bool _isValid = false;
+
+        /// <summary>
+        /// Creates an instance of <see cref="ValidationResult"/>
+        /// </summary>
+        protected ValidationResult()
+        {
+            ValidationFailureType = ValidationFailureType.ValidationNotEvaluated;
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="ValidationResult"/>
+        /// </summary>
+        /// <param name="validationFailureType">The <see cref="ValidationFailureType"/> that occurred during validation.</param>
+        protected ValidationResult(ValidationFailureType validationFailureType)
+        {
+            ValidationFailureType = validationFailureType;
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="ValidationResult"/>
+        /// </summary>
+        /// <param name="validationFailureType">The <see cref="ValidationFailureType"/> that occurred during validation.</param>
+        /// <param name="exceptionDetail"> The <see cref="ExceptionDetail"/> representing the <see cref="Exception"/> that occurred during validation.</param>
+        protected ValidationResult(ValidationFailureType validationFailureType, ExceptionDetail exceptionDetail)
+        {
+            ValidationFailureType = validationFailureType;
+            ExceptionDetail = exceptionDetail;
+        }
+
+        /// <summary>
+        /// Adds a new stack frame to the exception details.
+        /// </summary>
+        /// <param name="stackFrame"></param>
+        public void AddStackFrame(StackFrame stackFrame)
+        {
+            ExceptionDetail.StackFrames.Add(stackFrame);
+        }
+
+        /// <summary>
+        /// Gets the <see cref="Exception"/> that occurred during validation.
+        /// </summary>
+        public abstract Exception Exception { get; }
+
+        /// <summary>
+        /// Gets the <see cref="ExceptionDetail"/> that occurred during validation.
+        /// </summary>
+        public ExceptionDetail ExceptionDetail { get; }
+
+        /// <summary>
+        /// True if the token was successfully validated, false otherwise.
+        /// </summary>
+        public bool IsValid
+        {
+            get
+            {
+                HasValidOrExceptionWasRead = true;
+                return _isValid;
+            }
+            set
+            {
+                _isValid = value;
+            }
+        }
+
+        // TODO - HasValidOrExceptionWasRead, IsValid, Exception are temporary and will be removed when TokenValidationResult derives from ValidationResult.
+        /// <summary>
+        /// Gets or sets a boolean recording if IsValid or Exception was called.
+        /// </summary>
+        protected bool HasValidOrExceptionWasRead { get; set; }
```

Comment:
I find the name to be a bit confusing. It reads as "Has Valid" or "Exception read". But every result has "IsValid". The name doesn't make it clear that it is if either has been read.

Reply:
If we want to be overly clear of the details we could go with something along the lines of `WasEitherIsValidOrExeceptionRead`, but could we simply say something like `ResultHasBeenValidated` or `ResultWasChecked`?